server/aesdsocket.c

```diff
@@ -30,303 +30,567 @@
-	}
+                tempPtr = (char*)realloc(bufferAppend, memAllocSize* sizeof(char));
 
-    freeaddrinfo(sockAddrInfo);
+                if(tempPtr!= NULL)
+                    bufferAppend=tempPtr;
 
-	return_value = listen(fd[FD_SOCKET], MAX_CONNECTION);
-	if(return_value == -1)
-	{
-		printf("ERROR: failed to listen, exiting...\n");
-		exit (EXIT_FAILURE);
-	}
+            }
+            //load into buffer
+            memcpy(&bufferAppend[bufferSize], buffer, receiveReturnValue);				
+        }
 
-    // ********************* open write file
-	fd[FD_DATA] = open(FILE_OUT_PATH, O_CREAT | O_APPEND | O_RDWR, S_IRWXU);
-    if(fd[FD_DATA] == -1)
-	{
-		printf("ERROR: failed to open file, %s \n", FILE_OUT_PATH);
-		exit (EXIT_FAILURE);
-	}
+        bufferSize+=receiveReturnValue;
 
-	sigemptyset(&sigSetOne);
-	sigaddset(&sigSetOne, SIGINT);
-	sigaddset(&sigSetOne, SIGTERM);
+        newlinePosition = strchr(buffer, '\n');	
 
-    buffer_rx = (char *)malloc(BUFFER_SIZE*sizeof(char));
-    buffer_tx = (char *)malloc(BUFFER_SIZE*sizeof(char));
-	if(buffer_rx == NULL || buffer_tx == NULL)
-	{
-		printf("ERROR: failed to malloc buffers, exiting...\n");
-		exit (EXIT_FAILURE);
-	}	
+    //keep going until new line found
+    }while(newlinePosition == NULL); 
 
-    if(daemonFlag)
-	{
-		pid = fork();
 
-		if(pid == -1)
-			exit (EXIT_FAILURE);
+    //lock mutex while writing
+    pthread_mutex_lock(&mutex);
+    writeReturnValue = write(fd[FD_DATA], bufferAppend, bufferSize);
+    if(writeReturnValue == -1)
+    {
+        syslog(LOG_ERR, "ERROR: Failed to write to file...");
+        return NULL;
+    }
+    //unlock mutex after writing
+    pthread_mutex_unlock(&mutex);	
 
-		else if(pid != 0)
-			exit(0);
 
-		if(setsid() == -1)
-			exit (EXIT_FAILURE);
+    endPosition = lseek(fd[FD_DATA], 0, SEEK_END);
 
-		if(chdir("/") == -1)
-			exit (EXIT_FAILURE);
+    retVal = lseek(fd[FD_DATA], 0, SEEK_SET);
+    if(retVal == (off_t)FAILURE)
+    {
+        syslog(LOG_ERR, "ERROR: Failed lseek for end of file position");
+        return NULL;
+    }
+
+    //while not at the end of file
+    while(sendIndex != endPosition)
+    {
+        toSendSize=0;
+        
+        char* bufferAppend2 = (char*)malloc(MAXSIZE*sizeof(char));
+        if(bufferAppend2 == NULL)
+        {
+            syslog(LOG_ERR,"ERROR: Failed to malloc appending buffer...");
+            break;
+        }
 
-		open("/dev/null", O_RDWR);
+        //clear buffer
+        memset(bufferAppend2, '\0', MAXSIZE);
 
-		dup(0);
-		dup(0);
-	}
+        //get size of next line
+        nextLineSize = endPosition - position;
+
+        do
+        {
+            //lock mutex while reading
+            pthread_mutex_lock(&mutex);										
+            readReturnValue = read(fd[FD_DATA], bufferAppend2+toSendSize, sizeof(char));
+
+            if(readReturnValue == FAILURE)
+            {
+                    syslog(LOG_ERR,"ERROR: Failed to read...");
+                    return NULL;
```

Comment:
@cchoi22915 same issue with locking here... think about what happens when you `return` here.